nativelib/src/main/resources/gc/commix/Marker.c

```diff
@@ -0,0 +1,367 @@
+        end_ns = scalanative_nano_time();
+        Stats_RecordEvent(stats, event_sync, start_ns, end_ns);
+    }
+#endif
+}
+
+static inline void Marker_giveFullPacket(Heap *heap, Stats *stats, GreyPacket *packet) {
+    assert(packet->type == grey_packet_refrange || packet->size > 0);
+    // make all the contents visible to other threads
+    atomic_thread_fence(memory_order_acquire);
+    assert(GreyList_Size(&heap->mark.full) <= heap->mark.total);
+#ifdef ENABLE_GC_STATS_SYNC
+    uint64_t start_ns, end_ns;
+    if (stats != NULL) {
+        start_ns = scalanative_nano_time();
+    }
+#endif
+    GreyList_Push(&heap->mark.full, heap->greyPacketsStart, packet);
+#ifdef ENABLE_GC_STATS_SYNC
+    if (stats != NULL) {
+        end_ns = scalanative_nano_time();
+        Stats_RecordEvent(stats, event_sync, start_ns, end_ns);
+    }
+#endif
+}
+
+void Marker_markObject(Heap *heap, Stats *stats, GreyPacket **outHolder, Bytemap *bytemap,
+                       Object *object, ObjectMeta *objectMeta) {
+    assert(ObjectMeta_IsAllocated(objectMeta) || ObjectMeta_IsMarked(objectMeta));
+
+    assert(Object_Size(object) != 0);
+    Object_Mark(heap, object, objectMeta);
+
+    GreyPacket *out = *outHolder;
+    if (!GreyPacket_Push(out, object)) {
+        Marker_giveFullPacket(heap, stats, out);
+        *outHolder = out = Marker_takeEmptyPacket(heap, stats);
+        GreyPacket_Push(out, object);
+    }
+}
+
+void Marker_markConservative(Heap *heap, Stats *stats, GreyPacket **outHolder, word_t *address) {
+    assert(Heap_IsWordInHeap(heap, address));
+    Object *object = Object_GetUnmarkedObject(heap, address);
+    Bytemap *bytemap = heap->bytemap;
+    if (object != NULL) {
+        ObjectMeta *objectMeta = Bytemap_Get(bytemap, (word_t *)object);
+        assert(ObjectMeta_IsAllocated(objectMeta));
+        if (ObjectMeta_IsAllocated(objectMeta)) {
+            Marker_markObject(heap, stats, outHolder, bytemap, object, objectMeta);
+        }
+    }
+}
+
+void Marker_markRange(Heap *heap, Stats *stats, GreyPacket* in, GreyPacket **outHolder, Bytemap *bytemap,
+                      word_t **fields, size_t length) {
+    for (int i = 0; i < length; i++) {
+        word_t *field = fields[i];
+        if (Heap_IsWordInHeap(heap, field)) {
+            ObjectMeta *fieldMeta = Bytemap_Get(bytemap, field);
+            if (ObjectMeta_IsAllocated(fieldMeta)) {
+                Marker_markObject(heap, stats, outHolder, bytemap,
+                                  (Object *)field, fieldMeta);
+            }
+        }
+    }
+}
+
+void Marker_markPacket(Heap *heap, Stats *stats, GreyPacket* in, GreyPacket **outHolder) {
+    Bytemap *bytemap = heap->bytemap;
+    if (*outHolder == NULL) {
+        GreyPacket *fresh = Marker_takeEmptyPacket(heap, stats);
+        assert(fresh != NULL);
+        *outHolder = fresh;
+    }
+    while (!GreyPacket_IsEmpty(in)) {
+        Object *object = GreyPacket_Pop(in);
+
+        if (Object_IsArray(object)) {
```

Comment:
I'd move the two branches of this if into two helper methods `Marker_markArrayObject` and `Marker_markRegularObject`.